<h1 class="r-stretch" style="text-align:center"> Introduction to Python, Part II  </h1>

<center>
    <br>
     <table style="background-color: white;">
        <tr>
            <td>
    <img src="https://www.python.org/static/community_logos/python-logo.png" width=250 style="float:center">
            </td>
            <td>
    <img src="https://brand.jhu.edu/assets/uploads/sites/5/2014/06/university.logo_.small_.horizontal.blue_.jpg" width=250 style="float:center">
            </td>
            </tr>
    </table>
        <br>
    Johns Hopkins University  <br>
Krieger School of Arts and Sciences  <br>
Office of Undergraduate Research, Scholarly and Creative Activity (URSCA) <br>
May 24, 2022 <br>
    <br>
    Prof. Daniel Beller <br>
    JHU Department of Physics &amp; Astronomy
    </center>
 </section>

### Yesterday

We introduced Python's fundamentals for
* math: <span class="new_term">integers</span> and <span class="new_term">floats</span>
* text: <span class="new_term">strings</span>
* comparisons: (`>`, `==`, etc.)
* logic: `and`/`or`/`not`

Multi-line programs were run top-to-bottom. For example,

In [53]:
bill = 76.15            # runs 1st
my_bill = bill / 4      # runs 2nd
my_tip = my_bill * 0.18 # runs 3rd
                        # ... 

### Today

We will cover how to structure a Python program: <span class="new_term">"control flow"</span>

Different blocks of code will be run...
* ...in different orders 
* ...or not at all
* ...or repeatedly! 

depending on the input data.
<br><br><br>


## `if` statements

In words, `if` statements let us implement this kind of logic:

>if we have situation A: <br>
>&emsp;&emsp;    then run this block of code...

To check whether we *do* have "situation A", we usually use <span class="new_term">comparisons</span> to create a <span class="new_term">conditional</span>: an expression that evaluates to `True` or `False`.

For example:

In [2]:
age = 14

if age < 16.5:
    print("You are too young to drive.")
    years_until_can_drive = 16.5 - age
    print("Wait", years_until_can_drive, "more years.")

You are too young to drive.
Wait 2.5 more years.


Here, the <span class="new_term">conditional</span> is:

In [3]:
age < 16.5

True

Notice formatting:

* The `if` statement ends in colon `:`
* Following this is an <span class="new_term">indented code block</span> containing any number of lines. (Use 'tab' key.)

<span class="your_turn">Try it!</span> – What happens if we change the conditional to a `True` one? 
<br><br><br>

### `if`-`else` 

Often we want our program to do something *different* if the conditional is `False`. 

For this, we follow the `if` block with an `else` block.


In [56]:
age = 20

if age < 16.5:
    print("You are too young to drive.")
else:
    print("Here are the keys to your new car!")

Here are the keys to your new car!


### `if`-`elif`-`else`

When there are more than two possibilities to consider, we can put one or more `elif` blocks between the `if` block and the `else` block. Each block is only executed if the conditional above it was `False`.

In [57]:
age = 17

if age < 16.5:
    print("You are too young to drive.")
elif age < 18: # another conditional!
    print("You can drive, but you can't buy lottery tickets yet.")
else:
    print("You can buy a new car with your lottery winnings!")

You can drive, but you can't buy lottery tickets yet.


## `while` loops

What do we get if we add up all the integers $1, 2, \dots 100$?

It would be annoying if we had to type out all 100 numbers!

Fortunately, programming languages allow <span class="new_term">loops</span> that run a block of code repeatedly.

In [58]:
total = 0
n = 1

while n <= 100:         # loop until n > 100
    total = total + n   # add n to total        
    print(n, total)
    n = n + 1           # add 1 to n     

1 1
2 3
3 6
4 10
5 15
6 21
7 28
8 36
9 45
10 55
11 66
12 78
13 91
14 105
15 120
16 136
17 153
18 171
19 190
20 210
21 231
22 253
23 276
24 300
25 325
26 351
27 378
28 406
29 435
30 465
31 496
32 528
33 561
34 595
35 630
36 666
37 703
38 741
39 780
40 820
41 861
42 903
43 946
44 990
45 1035
46 1081
47 1128
48 1176
49 1225
50 1275
51 1326
52 1378
53 1431
54 1485
55 1540
56 1596
57 1653
58 1711
59 1770
60 1830
61 1891
62 1953
63 2016
64 2080
65 2145
66 2211
67 2278
68 2346
69 2415
70 2485
71 2556
72 2628
73 2701
74 2775
75 2850
76 2926
77 3003
78 3081
79 3160
80 3240
81 3321
82 3403
83 3486
84 3570
85 3655
86 3741
87 3828
88 3916
89 4005
90 4095
91 4186
92 4278
93 4371
94 4465
95 4560
96 4656
97 4753
98 4851
99 4950
100 5050


The answer is $5050$. <br><br>






## Functions 

<span class="new_term">Functions</span> are the most important concept for *reusable* programs. 

A <span class="new_term">function</span> is a rule for mapping inputs to outputs. 



This is similar to mathematical functions like 

$$ f(x) = 2x+1 $$ 

Here the input is $x$ and the output is $2x+1$. 

For example, an input of $3$ gets mapped to an output of $7$.

The equivalent function in Python looks like this:

In [59]:
def f(x):
    return 2 * x + 1

Important features:
* `def` makes `f` a function 
* input <span class="new_term">argument</span> goes inside parentheses `()`
* output follows `return` on same line
* colon and indent just like `if` blocks

Now we can use the function for *any* input number!

In [60]:
f(3)

7

In [61]:
f(-201)

-401

<span class="your_turn">Try it!</span> – Functions

Write a function that calculates a 6% Maryland sales tax on any price. 

Then use it to find the sales tax for a book marked \$12.99.

In [62]:
def sales_tax(price):
    amount = price * 0.06
    answer = round(amount, 2)
    return answer 

sales_tax(12.99)

0.78

<br><br><br>

## Lists and `for` loops

<span class="new_term">Lists</span> are one of Python's most convenient features. Make a list by putting any collection of values inside `[]` and separated by commas `,`.

In [63]:
my_list = [3, -201, 11.1]

A `for` loop is similar to a `while` loop, except that a `for` loop iterates over the contents of your list and then stops automatically.

In [64]:
for number in my_list: # for each item in my_list, labeled as "number"...
    print(2 * number)  # multiply "number" by 2 and show me the result

6
-402
22.2


Notice that `number` is just a placeholder, like $x$ in algebra. We'd get the same result with:

In [65]:
for banana in my_list:
    print(2 * banana)

6
-402
22.2


We can combine <span class="new_term">functions</span> with `for` loops. For example, using our `f(x)` from above:

In [66]:
for number in my_list:
    print(f(number))

7
-401
23.2


<br><br><br>

### `range`

<span class="fragment">In a `for` loop, if we have a sequence of numbers like</span>

In [67]:
for i in [1, 2, 3, 4]:
    print(f(i))

3
5
7
9


we can use a `range` instead:

In [68]:
for i in range(1, 5): # start at 1, stop BEFORE 5
    print(f(i))

3
5
7
9


Adding up all the integers $1, 2, \dots, 100$ is easier this way:

In [69]:
total = 0

for i in range(1, 101):
    total = total + i
    
print(total)

5050


Actually, Python gives us an even better shortcut with its `sum` function:

In [70]:
sum(range(1,101))

5050

<br><br><br>

### Working with strings in `for` loops

How many times do the Beatles say the word "hand" in the lyrics to "I Want to Hold Your Hand"?

First we can use the `.split()` method to chop up a string into a <span class="new_term">list</span> of one-word strings.

In [6]:
lyrics = """
Oh yeah, I'll tell you something
I think you'll understand
When I say that something
I wanna hold your hand
I wanna hold your hand
I wanna hold your hand

Oh please, say to me
You'll let me be your man
And please, say to me
You'll let me hold your hand
You'll let me hold your hand
I wanna hold your hand

And when I touch you I feel happy
Inside
It's such a feeling that my love
I can't hide
I can't hide
I can't hide

Yeah, you got that something
I think you'll understand
When I say that something
I wanna hold your hand
I wanna hold your hand
I wanna hold your hand

And when I touch you I feel happy
Inside
It's such a feeling that my love
I can't hide
I can't hide
I can't hide

Yeah, you got that something
I think you'll understand
When I feel that something
I wanna hold your hand
I wanna hold your hand
I wanna hold your hand
I wanna hold your hand
"""

In [7]:
words = lyrics.split()

print(words)

['Oh', 'yeah,', "I'll", 'tell', 'you', 'something', 'I', 'think', "you'll", 'understand', 'When', 'I', 'say', 'that', 'something', 'I', 'wanna', 'hold', 'your', 'hand', 'I', 'wanna', 'hold', 'your', 'hand', 'I', 'wanna', 'hold', 'your', 'hand', 'Oh', 'please,', 'say', 'to', 'me', "You'll", 'let', 'me', 'be', 'your', 'man', 'And', 'please,', 'say', 'to', 'me', "You'll", 'let', 'me', 'hold', 'your', 'hand', "You'll", 'let', 'me', 'hold', 'your', 'hand', 'I', 'wanna', 'hold', 'your', 'hand', 'And', 'when', 'I', 'touch', 'you', 'I', 'feel', 'happy', 'Inside', "It's", 'such', 'a', 'feeling', 'that', 'my', 'love', 'I', "can't", 'hide', 'I', "can't", 'hide', 'I', "can't", 'hide', 'Yeah,', 'you', 'got', 'that', 'something', 'I', 'think', "you'll", 'understand', 'When', 'I', 'say', 'that', 'something', 'I', 'wanna', 'hold', 'your', 'hand', 'I', 'wanna', 'hold', 'your', 'hand', 'I', 'wanna', 'hold', 'your', 'hand', 'And', 'when', 'I', 'touch', 'you', 'I', 'feel', 'happy', 'Inside', "It's", 'such

Then we can do a `for` loop over the elements of the list `words`, using an `if` statement to check if each word is "hand":

In [8]:
total = 0

for word in words:
    if word == "hand":  # string equality comparison 
        total = total + 1
        
print("The word 'hand' appears", total, "times in this song.")       

The word 'hand' appears 13 times in this song.


<br><br><br>

### List elements

What is the 24th word in the song? Use square brackets `[]` to access an element of a list.

In [9]:
words[23]

'your'

Why `23` and not `24`? Because **Python starts counting from zero, not one.**

In [10]:
words[0] # 1st word

'Oh'

In [11]:
words[1] # 2nd word

'yeah,'

How many words are in the song? 

In [12]:
len(words)

176

What are the first 10 words in the song? For this we can use a <span class="new_term">slice</span>

In [13]:
first_ten = words[0:10] # start at 1st word, stop before 11th word
print(first_ten)

['Oh', 'yeah,', "I'll", 'tell', 'you', 'something', 'I', 'think', "you'll", 'understand']


We can glue these words back together into a single string (with spaces in between the words) in this way:

In [14]:
" ".join(first_ten)

"Oh yeah, I'll tell you something I think you'll understand"

When does the word "something" first appear in the song?

In [15]:
words.index("something") 

5

It's word number 6! (Since index 0 is word number 1.)  <br><br><br>

## List comprehension 

You can construct lists from `for` loops in a single line. This is called a <span class="new_term">list comprehension</span>.

For example, to show the length of each word in the lyrics (in order):

In [16]:
word_lengths = [ len(word) for word in words ]
print(word_lengths)

[2, 5, 4, 4, 3, 9, 1, 5, 6, 10, 4, 1, 3, 4, 9, 1, 5, 4, 4, 4, 1, 5, 4, 4, 4, 1, 5, 4, 4, 4, 2, 7, 3, 2, 2, 6, 3, 2, 2, 4, 3, 3, 7, 3, 2, 2, 6, 3, 2, 4, 4, 4, 6, 3, 2, 4, 4, 4, 1, 5, 4, 4, 4, 3, 4, 1, 5, 3, 1, 4, 5, 6, 4, 4, 1, 7, 4, 2, 4, 1, 5, 4, 1, 5, 4, 1, 5, 4, 5, 3, 3, 4, 9, 1, 5, 6, 10, 4, 1, 3, 4, 9, 1, 5, 4, 4, 4, 1, 5, 4, 4, 4, 1, 5, 4, 4, 4, 3, 4, 1, 5, 3, 1, 4, 5, 6, 4, 4, 1, 7, 4, 2, 4, 1, 5, 4, 1, 5, 4, 1, 5, 4, 5, 3, 3, 4, 9, 1, 5, 6, 10, 4, 1, 4, 4, 9, 1, 5, 4, 4, 4, 1, 5, 4, 4, 4, 1, 5, 4, 4, 4, 1, 5, 4, 4, 4]


How many characters does the longest word have? 

In [17]:
max(word_lengths)

10

<h3 class="fragment semi-fade-out">List comprehension </h3>

We can put an `if` statement inside a list comprehension to filter the list before it goes through the `for` loop. 

For example, to get all the words in `lyrics` that start with "y":

In [18]:
my_list = [ word for word in words if word.startswith("y") ]
print(my_list)

['yeah,', 'you', "you'll", 'your', 'your', 'your', 'your', 'your', 'your', 'your', 'you', 'you', "you'll", 'your', 'your', 'your', 'you', 'you', "you'll", 'your', 'your', 'your', 'your']


This is a shortcut for the following `for` loop with `if` block inside it:

In [19]:
my_list = []

for word in words:
    if word.startswith("y"):       
        my_list.append(word)
        
print(my_list)        
        

['yeah,', 'you', "you'll", 'your', 'your', 'your', 'your', 'your', 'your', 'your', 'you', 'you', "you'll", 'your', 'your', 'your', 'you', 'you', "you'll", 'your', 'your', 'your', 'your']


<span class="your_turn">Try it yourself!</span> — In your song lyrics from yesterday, find all the words whose first letter is "h".

Can you use `len()` to get the *number* of these words? 

In [22]:
words_starting_with_D = [ word for word in lyrics.split() if word.startswith("h") ]
print(words_starting_with_D)
print(len(words_starting_with_D))

['hold', 'hand', 'hold', 'hand', 'hold', 'hand', 'hold', 'hand', 'hold', 'hand', 'hold', 'hand', 'happy', 'hide', 'hide', 'hide', 'hold', 'hand', 'hold', 'hand', 'hold', 'hand', 'happy', 'hide', 'hide', 'hide', 'hold', 'hand', 'hold', 'hand', 'hold', 'hand', 'hold', 'hand']
34


<br><br><br>

## Summary

* `if`-`elif`-`else` statements allow your program to make "decisions"
* `while` loops, `for` loops, and <span class="new_term">functions</span> let you use the same block of code repeatedly.
* <span class="new_term">Lists</span> make it easy to work with data such as a collection of numbers or long text.

### Next steps:

You can download Python from www.python.org and run your own programs on your computer. 



There are many free resources to help you learn Python at you own pace. 

<ul style="margin-top:-0px;">
    <li>"Automate the Boring Stuff with Python" <a href="https://automatetheboringstuff.com">automatetheboringstuff.com</a>, a beginner's guide to Python scripting for everyday use </li>
    <li> Many other options at <a href="https://wiki.python.org">wiki.python.org</a></li>
    </ul>

We didn't have time to cover some of the most fun and useful elements of Python, such as <span class="new_term">classes</span> and <span class="new_term">modules</span>... look these up next! 

These slides will be posted and linked from [bellerphysics.gitlab.io/Code.html](https://bellerphysics.gitlab.io/Code.html) — which is a webpage created using Python!

Happy coding!

<br><br><br>